text/JanHavranek/lcp_lr_opis.tex

```diff
@@ -0,0 +1,91 @@
+\documentclass[12pt]{article}
+\usepackage[T1]{fontenc}
+\usepackage[utf8]{inputenc}
+\usepackage{amsmath}
+\usepackage{amsthm}
+
+\newtheorem*{theorem}{Twierdzenie}
+
+\title{Manber, Myers: zapytania do tablicy sufiksowej w czasie $O(m + \log{} n)$}
+\author{Jan Havránek}
+\date{20.06.2020}
+
+
+\begin{document}
+
+\maketitle
+
+\section*{Wprowadzenie}
+Tablica sufiksowa $SA$ zawiera początkowe indeksy leksykograficznie uporządkowanych sufiksów słowa $t$. Oznaczmy taki sufiks wskazany przez indeks $a$ w $SA$ jako $t_a = t[SA[a]\ldots n]$. Mając tablicę sufiksową, można ją wykorzystać do indeksowania tekstu -- poprzez zwykłe wyszukiwanie binarne można odnaleźć pierwszą ($l_w$) i ostatnią ($r_w$) pozycję w $SA$ taką, że $w$ jest prefiksem $t_i, i \in l_w \leq i \leq r_w$. Wtedy $SA[i]$ wskazują wszystkie wystąpienia $w$ w $t$. Jednak to podejście wymaga $O(m \log{} n)$ czasu.
+
+Manber i Myers w 1993 r. zaproponowali prosty algorytm przyśpieszający takie zapytania do $O(m + \log{} n)$. Wykorzystywana jest do tego struktura $LCP\textrm{-}LR$ zawierająca najdłuższe wspólne prefiksy (\textit{lcp}) sufiksów $w$. 
+
+\section*{Zapytania do $SA$ z wykorzystaniem $LCP\textrm{-}LR$}
+
+Załóżmy, że mamy tablicę $SA$ oraz $LCP\textrm{-}LR$ (jej konstrukcja zostanie omówiona w następnej sekcji). Informację o \textit{lcp} sufiksów $t$ można wykorzystać do przyśpieszenia wyszukiwania binarnego, ograniczając liczbę wykonywanych porównań symboli. 
+
+Oznaczmy jako $l$ i $r$ lewą i prawą granicę rozpatrywanego przedziału $SA$ w wyszukiwaniu binarnym. $c = (l + r) / 2 $ jest zatem pozycja oznaczająca sufiks $t_c$, z którym porównywamy wyszukiwane słowo $w$. Dalej oznaczmy $L = lcp(t_l, w)$ i $R = lcp(t_r, w)$. Podczas biegu algorytmu będą wartości $L$ i $R$ aktualizowane (niezmiennik), i to z wykorzystaniem minimalnej ilości porównań. Niech $H = \max (L, R)$ i $C = lcp(t_c, w)$. 
+
+Bez straty ogólności załóżmy, że w danym cyklu wyszukiwania $H = L \geq R$. Mogą zajść trzy przypadki (w praktyce można pierwsze dwa implementować razem):
+
+\begin{enumerate}
+    \item $lcp(t_l, t_m) > L$ \\
+          Wtedy $C = L$, skoro zgodność prefiksów jest relacją tranzytywną.
```

Comment:
Well, could you expand this by showing exactly how transitivity implies $C = L$? It's not at all obvious.